File: clavier_propre/service.py
"""Mode « service » (sans interface graphique).

Utilisé lorsque l'application est lancée avec l'argument ``--service``. Elle
applique la protection au démarrage puis s'exécute en boucle, en réappliquant
périodiquement la protection si elle a été désactivée par un tiers.

C'est cette fonction qui sert de point d'entrée à l'exécutable Windows
installé comme service (via NSSM ou pywin32 ``win32serviceutil``).
"""
from __future__ import annotations

import logging
import signal
import threading

from clavier_propre.controller import ProtectionController

log = logging.getLogger(__name__)

_STOP = threading.Event()


def _loop(controller: ProtectionController) -> None:
    controller.apply(active=True)
    while not _STOP.wait(30):
        st = controller.current_state()
        if controller.config.protection_active and st.keyboard_enabled:
            log.info("Réactivation détectée : re-protection (mode service).")
            controller.apply(active=True)


def run_headless() -> int:
    controller = ProtectionController()

    def stop_handler(signum, _frame):
        log.info("Arrêt du service demandé (signal %s).", signum)
        _STOP.set()

    signal.signal(signal.SIGINT, stop_handler)
    signal.signal(signal.SIGTERM, stop_handler)

    log.info("Démarrage du service Clavier-Propre (mode headless).")
    _loop(controller)
    log.info("Service arrêté.")
    return 0

File: clavier-rs/src/service.rs
//! Mode « service » (sans interface graphique).
//!
//! Lancé avec l'argument `--service`, l'application applique la protection au
//! démarrage puis s'exécute en boucle, en réappliquant périodiquement la
//! protection si elle a été désactivée par un tiers.
//!
//! Pour l'installer en tant que service Windows natif, on fournit un script
//! NSSM (voir `service/install-nssm.bat`). Rust ne peut pas facilement déclarer
//! un service natif sans dépendances C lourdes ; NSSM est la solution la plus
//! simple et la plus robuste pour un exécutable autonome.

use crate::controller::ProtectionController;
use std::sync::Arc;
use std::time::Duration;

pub fn run_headless() -> i32 {
    let controller = Arc::new(ProtectionController::new());
    controller.apply(true);
    println!("Service Clavier-Propre démarré (mode headless).");

    loop {
        std::thread::sleep(Duration::from_secs(30));
        controller.watchdog();
    }
}
